src/needless_continue.rs

```diff
@@ -0,0 +1,547 @@
+                                             if_cond: &Expr,
+                                             else_expr: &Expr,
+                                             header: &str) -> String {
+
+    let cond_code = &pprust::expr_to_string(&simplify(&negate_expr(if_cond)));
+    let if_code = &format!("if {} {{", cond_code);
+
+    let span = Span { lo: else_expr.span.lo + BytePos::from_usize(1usize),
+                      hi: else_expr.span.hi,
+                      expn_id: else_expr.span.expn_id };
+    let else_code = &trim_indent(&snippet(ctx, span, "..").into_owned(), false);
+
+    // Since we do not really know how much to indent here, we'll just indent
+    // the if block with 4 spaces.
+    let mut ret = String::from(header);
+    ret.push_str(if_code);
+    ret.push_str(&left_pad_with_spaces(&trim_indent(else_code, false), 4usize));
+    ret
+}
+
+/// Created a Spanned<T> with a dummy Span.
+fn mk_spanned<T>(node: T) -> Spanned<T> {
+    Spanned {
+        node: node,
+        span: dummy_span()
+    }
+}
+
+/// Creates a dummy Span with all fields set to zero.
+fn dummy_span() -> Span {
+    Span { lo: BytePos::from_usize(0usize),
+           hi: BytePos::from_usize(0usize),
+           expn_id: ExpnId::from_u32(0u32) }
+}
+
+/// Returns the dual comparison operator (under boolean negation) of `op`.
+fn dual_comparison_operator(op: BinOp_) -> Option<BinOp_> {
+    match op {
+        BiLt  => Some(BiGe),
+        BiLe  => Some(BiGt),
+        BiGt  => Some(BiLe),
+        BiGe  => Some(BiLt),
+        BiEq  => Some(BiNe),
+        BiNe  => Some(BiEq),
+        _     => None,
+    }
+}
+
+/// Returns the de Morgan's dual operator of `op` if it is either `&&` or `||`.
+fn dual_boolean_operator(op: BinOp_) -> Option<BinOp_> {
+    match op {
+        BiAnd => Some(BiOr),
+        BiOr  => Some(BiAnd),
+        _     => None,
+    }
+}
+
+fn is_comparison_operator(op: BinOp_) -> bool {
+    match op {
+        BiLt|BiLe|BiGt|BiGe|BiEq|BiNe => true,
+        _  => false,
+    }
+}
+
+fn is_logical_operator(op: BinOp_) -> bool {
+    match op {
+        BiOr|BiAnd => true,
+        _ => false,
+    }
+}
+
+// TODO: The following 3 functions that transform Exprs clone the exprs, since
+// syntax::ptr::P needs ownership of its inmate. Maybe there is a way around
+// these clones.
+
+/// Returns an expression representing the logical negation (unary not) of
+/// `expr`. If `expr` is already a unary not expression, the inner expression
+/// is returned (!!(x) being the same as x).
+fn negate_expr(expr: &Expr) -> Expr {
+    if let ExprUnary(UnNot, ref negated) = expr.node {
+        // !!x == x
+        // `negated` is of the type &P<Expr>, and hence the double indirection.
+        (**negated).clone()
+    } else {
+        Expr {
+            id: 0u32,
+            node: ExprUnary(
+                UnNot,
+                P((*expr).clone()),
+            ),
+            span: Span { lo: BytePos::from_usize(0usize),
```

Comment:
I'm also wary of constructing suggestion ASTs by hand. This again has bad interactions with macros.